### ISSUE TITLE
DPO loss mask computation

### ISSUE THREAD

**Issue by zkshan2002** [OpenRLHF/OpenRLHF#465]
Here is a piece of function `_get_batch_logps` of class `DPOTrainer` in `openrlhf/trainer/dpo_trainer.py`:
```
labels = labels[:, 1:].clone()
logits = logits[:, :-1, :]

loss_masks = attention_mask.clone().bool()
# mask prompts
for mask, source_len in zip(loss_masks, prompt_id_lens):
    mask[:source_len] = False
loss_masks = loss_masks[:, 1:]

# dummy token; we'll ignore the losses on these tokens later
labels[loss_masks == False] = 0
per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)).squeeze(2)

logprobs_sums = (per_token_logps * loss_masks).sum(-1)
```
Should it be `loss_masks = loss_masks [:, :-1]` instead?
To be clear, let's consider a toy example when `prompt_id_len=0`, the first token is pad_token and should not compute loss, i.e. `attention_mask[0] = 0`. After the shift, it should be `loss_masks[0] = 0` and `labels[0] = 0` dummy value. But in the current code, `attention_mask[0]` doesn't decide anything.

**Comment by hijkzzz** [comment]
prompt_id_len >= 1

**Comment by zkshan2002** [comment]
This is not my point, but I double checked and find that loss mask should indeed be shifted left. I'll close this issue.
